lua/weapons/weapon_magicsword.lua

```diff
@@ -0,0 +1,149 @@
+SWEP.PrimaryDelay = 0.38
+SWEP.PrimaryRange = 70
+
+SWEP.HitSound = Sound("weapons/blades/swordchop.mp3")
+SWEP.SwingSound = Sound("weapons/blades/woosh.mp3")
+SWEP.HitWallSound = Sound("weapons/blades/hitwall.mp3")
+
+if SERVER then
+	resource.AddFile("materials/magicsword/sword.vmt")
+end
+
+function SWEP:Initialize()
+	self:SetWeaponHoldType(self.HoldType)
+	self.ViewModelFOV = 75
+end
+
+function SWEP:Think()
+	if self.Idle and CurTime() >= self.Idle then
+		self.Idle = nil
+		self:SendWeaponAnim(ACT_VM_IDLE)
+	end
+end
+
+function SWEP:PrimaryAttack()
+	local tr = {}
+	tr.start = self:GetOwner():GetShootPos()
+	tr.endpos = self:GetOwner():GetShootPos() + (self:GetOwner():GetAimVector() * self.PrimaryRange)
+	tr.filter = self:GetOwner()
+	tr.mask = MASK_SHOT_HULL
+	local trace = util.TraceLine(tr)
+
+	self:SetNextPrimaryFire(CurTime() + self.PrimaryDelay)
+	self:GetOwner():SetAnimation(PLAYER_ATTACK1)
+
+	if trace.Hit then
+		if trace.Entity:IsPlayer() or trace.Entity:IsNPC() then
+			self:SetNextPrimaryFire(CurTime() + self.PrimaryDelay)
+			self:SendWeaponAnim(self.HitAnim)
+			self.Idle = CurTime() + self:GetOwner():GetViewModel():SequenceDuration()
+			bullet = {}
+			bullet.Num = 1
+			bullet.Src = self:GetOwner():GetShootPos()
+			bullet.Dir = self:GetOwner():GetAimVector()
+			bullet.Spread = Vector(0, 0, 0)
+			bullet.Tracer = 0
+			bullet.Damage = self.PrimaryDamage
+			self:GetOwner():FireBullets(bullet)
+			self:EmitSound(self.HitSound)
+		else
+			bullet = {}
+			bullet.Num = 1
+			bullet.Src = self:GetOwner():GetShootPos()
+			bullet.Dir = self:GetOwner():GetAimVector()
+			bullet.Spread = Vector(0, 0, 0)
+			bullet.Tracer = 0
+			bullet.Damage = self.PrimaryDamage
+			self:GetOwner():FireBullets(bullet)
+			self:SetNextPrimaryFire(CurTime() + self.PrimaryDelay)
+			self:SendWeaponAnim(self.HitAnim)
+			self.Idle = CurTime() + self:GetOwner():GetViewModel():SequenceDuration()
+			self:EmitSound(self.HitWallSound)
+		end
+	else
+		self:EmitSound(self.SwingSound)
+		self:SendWeaponAnim(self.MissAnim)
+		self.Idle = CurTime() + self:GetOwner():GetViewModel():SequenceDuration()
+	end
+end
+
+function SWEP:SecondaryAttack()
+	self:SetNextSecondaryFire(CurTime() + 2)
+	local owner = self:GetOwner()
+	if IsValid(owner) and owner:IsOnGround() then
+		local looking = owner:GetAimVector()
+		local velocity = looking * 1000
+
+		owner:SetVelocity(velocity)
+		owner:EmitSound("ambient/levels/labs/electric_explosion1.wav")
+	end
+end
+
+function SWEP:Reload()
+	local owner = self:GetOwner()
+	if CurTime() - (self.LastHealTime or 0) >= 2 and owner:Health() < 100 then
+		local currentHealth = owner:Health()
+		local maxHealth = owner:GetMaxHealth()
+		local healAmount = math.min(maxHealth - currentHealth, 20)
+
+		owner:SetHealth(currentHealth + healAmount)
+		self.LastHealTime = CurTime()
+		owner:EmitSound("items/smallmedkit1.wav")
+	end
+end
+
+hook.Add("EntityTakeDamage", "PreventFallDamage", function(target, dmginfo)
```

Comment:
It looks like the [`GetFallDamage` hook](https://wiki.facepunch.com/gmod/GM:GetFallDamage) would be the better choice here